### PR TITLE
add userdata param to callback function

### PR DESCRIPTION
**Pull request by zydxhs**
Add UserData to the callback function to satisfy concurrent calls in a multi-threaded environment, and put the userdata parameter last.

### CODE REVIEW

File: test/callback_tests.cpp [hunk 1]
```diff
@@ -863,15 +863,15 @@ TEST(CallbackDataTests, CallbackWriteFunctionTextTest) {
 
 TEST(CallbackDataTests, CallbackProgressFunctionCancelTest) {
     Url url{server->GetBaseUrl() + "/url_post.html"};
-    Response response = Post(url, ProgressCallback{[](size_t /*downloadTotal*/, size_t /*downloadNow*/, size_t /*uploadTotal*/, size_t /*uploadNow*/) -> bool { return false; }});
+    Response response = Post(url, ProgressCallback{[](intptr_t /*userdata*/, size_t /*downloadTotal*/, size_t /*downloadNow*/, size_t /*uploadTotal*/, size_t /*uploadNow*/) -> bool { return false; }});
```
Comment (COM8) [inline test/callback_tests.cpp:866]:
```suggestion
    Response response = Post(url, ProgressCallback{[](size_t /*downloadTotal*/, size_t /*downloadNow*/, size_t /*uploadTotal*/, size_t /*uploadNow*/, intptr_t /*userdata*/) -> bool { return false; }});
```

Comment (COM8) [reply test/callback_tests.cpp:866]: Probably the wrong order.

File: test/callback_tests.cpp [hunk 2]
```diff
@@ -863,15 +863,15 @@ TEST(CallbackDataTests, CallbackWriteFunctionTextTest) {
...
     EXPECT_EQ(response.error.code, ErrorCode::REQUEST_CANCELLED);
 }
 
 TEST(CallbackDataTests, CallbackProgressFunctionTotalTest) {
     Url url{server->GetBaseUrl() + "/url_post.html"};
     Body body{"x=5"};
     size_t response_upload = 0, response_download = 0;
-    Response response = Post(url, body, ProgressCallback{[&](size_t downloadTotal, size_t /*downloadNow*/, size_t uploadTotal, size_t /*uploadNow*/) -> bool {
+    Response response = Post(url, body, ProgressCallback{[&](intptr_t /*userdata*/, size_t downloadTotal, size_t /*downloadNow*/, size_t uploadTotal, size_t /*uploadNow*/) -> bool {
```
Comment (COM8) [inline test/callback_tests.cpp:874]:
```suggestion
    Response response = Post(url, body, ProgressCallback{[&](size_t downloadTotal, size_t /*downloadNow*/, size_t uploadTotal, size_t /*uploadNow*/, intptr_t /*userdata*/) -> bool {
```

**Review by COM8 — CHANGES_REQUESTED** [review]
Looks good. Only those two small variable reordering changes required.

**Comment by COM8** [comment]
Docs: https://github.com/whoshuu/cpr/pull/600

**Comment by zydxhs** [comment]
I'm really sorry for the extra work caused by my mistake.

I have fixed the error.

**Comment by zydxhs** [comment]
I'm really sorry for the extra work caused by my mistake.

I have fixed the error.

**Review by COM8 — APPROVED** [review]

**Comment by COM8** [comment]
Thanks for your time with this PR!

**Comment by Chronial** [comment]
I am somewhat confused by this API-Breaking change: What is the point of the userdata parameter? The callback is a std::function and can thus contain arbitrary data / state.

What is the difference between `cpr::ReadCallback{-1, [userdata = 23](char*, size_t& ) -> size_t { ... }}` 
and the new infertace `cpr::ReadCallback{-1, [](char*, size_t&, intptr_t userdata) -> size_t { ... }, 23}` ?

**Comment by COM8** [comment]
Hmmm, I have to agree. This should have been solvable without this additional parameter. @zydxhs is there any specific use case, which prevents your use case from using the `std::function` capture?

**Pull request merged**